### PR TITLE
MM-10516: Added support for PostActions in ephemeral posts

### PR DESCRIPTION
**Pull request by levb**
#### Summary

The general approach is that we take all the metadata that `DoPostAction` needs to process client `model.DoPostActionRequests`, and store it in a serialized, encrypted `Cookie` field, in the `model.PostAction` struct. The client then must send it back, and it is then used to process `PostActions` as a fallback top the metadata in the database.

The secret for encrypting the cookie is generated and stored similarly to AsymmetricSigningKey.

- model
  - Added type `model.PostActionCookie`, and a `Cookie string` to `model.PostAction`.
  - Added `model.AddPostActionCookies()`, `model.encryptActionCookie()`, `model.decryptActionCookie()`.
  - Added `Client4.DoPostActionWithCookie()`.
- api4
  - Changed `doPostAction` to use `ChannelId` from `PostActionCookie` when available for permission check 
  - Changed `createEphemeralPost` add `PostActionCookie` to result
- app
  - Added `App.ensurePostActionCookieSecret()`, `App.PostActionCookieSecret()`
  - Added `App.DoPostActionWithCookie()` to process incoming requests with
  cookies. For backward compatibility, it prefers the metadata in the
  database; falls back to cookie.
  - Changed `App.SendEphemeralPost()` to add PostActionCookie to outgoing websocket message
  - Added `plugin.API.UpdateEphemeralPost()` and `plugin.API.DeleteEphemeralPost()`.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10516

#### Checklist
- [X] Added or updated unit tests (required for all new features)

See also:
(a) https://github.com/mattermost/mattermost-redux/pull/769
(b) https://github.com/mattermost/mattermost-webapp/pull/2351
(c) https://github.com/levb/mattermost-giphy-plugin

### CODE REVIEW

File: app/post.go [hunk 1]
```diff
@@ -402,13 +408,160 @@ func (a *App) SendEphemeralPost(userId string, post *model.Post) *model.Post {
...
+
+	post.UpdateAt = model.GetMillis()
+	if post.Props == nil {
+		post.Props = model.StringInterface{}
+	}
+
+	post.GenerateActionIds()
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_POST_EDITED, "", post.ChannelId, userId, nil)
+
+	// TODO: why? but it doesn't work without clearing ChannelId, which is needed ^^.
```
Comment (levb) [inline app/post.go]: I could not find an explanation for this, but if ChannelId is set, something prevents the client from working right.

File: model/integration_action.go [hunk 2]
```diff
@@ -25,15 +25,49 @@ const (
...
+// PostAction.Cookie. The clients should hold on to it, and include it with
+// subsequent DoPostAction requests.  This allows the server to access the
+// action metadata even when it's not available in the database, for ephemeral
+// posts.
+type PostActionCookie struct {
+	Type        string                 `json:"type,omitempty"`
+	PostId      string                 `json:"post_id,omitempty"`
+	RootPostId  string                 `json:"root_post_id,omitempty"`
+	ChannelId   string                 `json:"channel_id,omitempty"`
+	TeamId      string                 `json:"team_id,omitempty"`
```
Comment (levb) [inline model/integration_action.go]: `TeamId` is probably unnecessary and irrelevant, _even_ if it were set correctly. Will double-check and remove.

File: model/config.go [hunk 3]
```diff
@@ -227,6 +229,7 @@ type ServiceSettings struct {
 	MaximumLoginAttempts                              *int
 	GoroutineHealthThreshold                          *int
 	GoogleDeveloperKey                                *string
+	ActionCookieSecret                                *string
```
Comment (jasonblais) [inline model/config.go]: Can you help describe the purpose of this setting - is it for security reasons? Is this absolutely necessary for admins to configure?

Comment (jasonblais) [reply model/config.go]: If we add this setting, we should 
- add its default value to config/default.json
- add if the default value is used or not to app/diagnostics.go

Comment (levb) [reply model/config.go]: It is the encryption key used to encrypt the cookies (plugin's Integration Context), specifically and hide them from the client. I does auto-default to a random value, but in a clustered config it should be set the same across all servers.

I'm 2/5 that tracking the default/override usage for this is not useful, would recommend against it. It should be unique for every cluster.

Comment (jasonblais) [reply model/config.go]: > I does auto-default to a random value

That means if someone upgrades to a release this feature ships in, and attempts to use PostActions in an ephemeral message, it will work out-of-the-box, and not require admin to configure/set the encryption key?

Comment (levb) [reply model/config.go]: As far as I understand, correct, https://github.com/mattermost/mattermost-server/pull/10258/files#diff-2279f86634da3f787f279961af76713aR320 takes care of it.

Comment (jasonblais) [reply model/config.go]: Okay. One more question - any objection for adding this setting in System Console UI, say in System Console > Integrations > Custom Integrations?

It would give the admin an ability to regenerate a new secret easily, like for public links (see below)
image

Comment (levb) [reply model/config.go]: Thank you this is exactly the feedback I needed (need to add this to console and docs)! Just need to confirm that this is the right thing in a clustered environment to have the value shared to all servers.

File: app/integration_action.go [hunk 4]
```diff
@@ -31,90 +31,152 @@ import (
...
+		if action == nil || action.Integration == nil {
+			return "", model.NewAppError("DoPostAction", "api.post.do_action.action_id.app_error", nil, fmt.Sprintf("action=%v", action), http.StatusNotFound)
+		}
+
+		upstreamRequest.ChannelId = post.ChannelId
+		upstreamRequest.TeamId = channel.TeamId
+		upstreamRequest.Type = action.Type
+		upstreamRequest.Context = action.Integration.Context
+
+		if action.Type == model.POST_ACTION_TYPE_SELECT {
```
Comment (levb) [inline app/integration_action.go]: I copied the condition from prior code, but it seems unnecessary, 3/5 the fields should be initialized unconditionally, and the values just not be relevant for non-select types. Seems to me that the code here should avoid any knowledge of the action type semantics.

File: app/integration_action.go [hunk 5]
```diff
@@ -31,90 +31,152 @@ import (
...
+		PostId: postId,
+	}
+
+	// See if the post exists in the DB, if so ignore the cookie.
+	// Start all queries here for parallel execution
 	pchan := a.Srv.Store.Post().GetSingle(postId)
 	cchan := a.Srv.Store.Channel().GetForPost(postId)
-
 	result := <-pchan
 	if result.Err != nil {
```
Comment (levb) [inline app/integration_action.go]: (0/5) We should consider defaulting to the Cookie when available, and set it on all posts (with actions) sent to clients. It should be backwards-compatible. This would save 2 queries/DoPostAction, and with a few more changes may eventually eliminate the need to store message-specific PostActions in the database altogether.

File: model/config.go (hunk 3 above)
Comment (hanzei) [reply model/config.go]: The server has a `asymmetricSigningKey` which is a `ecdsa.PrivateKey`(https://github.com/mattermost/mattermost-server/blob/master/app/server.go#L100). Maybe this can be used to encrypt the cookie? I only know about the existence, Christopher will have more insights in this.

File: app/integration_action.go (hunk 5 above)
Comment (crspeller) [reply app/integration_action.go]: Agreed.

File: app/integration_action.go [hunk 6]
```diff
@@ -31,90 +31,152 @@ import (
...
+			return "", result.Err
+		}
 
-	result = <-cchan
-	if result.Err != nil {
-		return "", result.Err
-	}
-	channel := result.Data.(*model.Channel)
+		// Get action metadata from the cookie
+		c := model.PostActionCookie{}
```
Comment (crspeller) [inline app/integration_action.go]: Single letter variable `c` sounds like context in the MM app. Can we maybe rename this one to `cookie` and the current cookie to cookieJSON or something?

File: app/integration_action.go (hunk 4 above)
Comment (crspeller) [reply app/integration_action.go]: +1

File: app/post.go [hunk 7]
```diff
@@ -402,13 +408,160 @@ func (a *App) SendEphemeralPost(userId string, post *model.Post) *model.Post {
...
 
 	return post
 }
 
+func (a *App) AddActionCookiesToPost(o *model.Post) *model.Post {
+	p := o.Clone()
+
+	// retainedProps carry over their value from the old post, including no value
+	retainPropKeys := []string{"override_username", "override_icon_url"}
+	retainProps := map[string]interface{}{}
```
Comment (crspeller) [inline app/post.go]: Can I get an explanation of what this "retain and remove props" stuff is?

**Comment by hanzei** [comment]
I tried this locally with Matterpoll and run into an error:
When clicking on a message button, nothing happens in the UI and I see the following message in the server log:
```bash
2019-02-13T11:58:52.170+0100	error	web/context.go:52	You do not have the appropriate permissions	{"path": "/api/v4/posts/rn1wohsjwjyzbmw84g3znof5yy/actions/nfiijxkidtgkbetehu7e9imeuo", "request_id": "oe49oxbnm3nou86cqi4eqi5n8e", "ip_addr": "127.0.0.1", "user_id": "ik565zsjnffm78wau8tjyajt3o", "method": "POST", "err_where": "Permissions", "http_code": 403, "err_details": "userId=ik565zsjnffm78wau8tjyajt3o, permission=read_channel"}
```
I'm 90% sure that this error is generated in https://github.com/mattermost/mattermost-server/blob/master/api4/integration_action.go#L26 when calling [`(a *App) SessionHasPermissionToChannelByPost`](https://github.com/mattermost/mattermost-server/blob/master/app/authorization.go#L65). I didn't look further into this, but it seams additional changes to this code part are needed.

File: model/integration_action.go [hunk 8]
```diff
@@ -25,15 +25,49 @@ const (
...
-	Options     []*PostActionOptions   `json:"options"`
+	// A unique Action ID. If not set, generated automatically.
+	Id string `json:"id"`
+
+	// The type of the interactive element. Currently supported are
+	// "select" and "button".
+	Type string `json:"type"`
+
+	// The text on the button, or in the select placeholder.
+	Name string `json:"name"`
```
Comment (hanzei) [inline model/integration_action.go]: Maybe use `omitempty`? We discussed this briefly in the dev meeting and didn't found any downsides.

Comment (hanzei) [inline model/integration_action.go]: Maybe use `omitempty`? We discussed this briefly in the dev meeting and didn't found any downsides.

File: app/post.go [hunk 9]
```diff
@@ -402,13 +408,160 @@ func (a *App) SendEphemeralPost(userId string, post *model.Post) *model.Post {
...
+
+	return post
+}
+
+func (a *App) UpdateEphemeralPost(userId string, post *model.Post) *model.Post {
+	post.Type = model.POST_EPHEMERAL
+
+	post.UpdateAt = model.GetMillis()
+	if post.Props == nil {
+		post.Props = model.StringInterface{}
```
Comment (hanzei) [inline app/post.go]: Why is this needed?

File: model/integration_action.go [hunk 10]
```diff
@@ -25,15 +25,49 @@ const (
...
+	// are "users" and "channels".
+	DataSource string               `json:"data_source"`
+	Options    []*PostActionOptions `json:"options"`
+
+	// Defines the interaction with the backend upon a user action.
+	// Integration contains Context, which is private plugin data;
+	// Integrations are stripped from Posts when they are sent to the
+	// client, or are encrypted in a Cookie.
 	Integration *PostActionIntegration `json:"integration,omitempty"`
+	Cookie      string                 `json:"cookie,omitempty" db:"-"`
```
Comment (hanzei) [inline model/integration_action.go]: What does `db:"-"` do? I've never seen this before.

File: plugin/api.go [hunk 11]
```diff
@@ -272,6 +272,12 @@ type API interface {
 	// SendEphemeralPost creates an ephemeral post.
 	SendEphemeralPost(userId string, post *model.Post) *model.Post
 
+	// UpdateEphemeralPost updates an ephemeral message previously sent to the user.
+	UpdateEphemeralPost(userId string, post *model.Post) *model.Post
```
Comment (hanzei) [inline plugin/api.go]: I hesitate to add a new API method for this. Why can't [`UpdatePost()`](https://developers.mattermost.com/extend/plugins/server/reference/#API.UpdatePost) be used?

File: plugin/api.go [hunk 12]
```diff
@@ -272,6 +272,12 @@ type API interface {
...
+
+	// DeleteEphemeralPost deletes an ephemeral message previously sent to the user.
+	DeleteEphemeralPost(userId string, post *model.Post)
```
Comment (hanzei) [inline plugin/api.go]: Same questions as for `UpdateEphemeralPost()`: Why do we need this?

**Comment by scottleedavis** [comment]
Also ran it locally, and was happy to see ephemeral buttons!   (very cool)
ephemeral buttons

File: app/post.go (hunk 7 above)
Comment (levb) [reply app/post.go]: This was in the code before, I just added it to the cookie use-case. As I understand it, "override_username", "override_icon_url" values including nils are carried from the original post to the updated one, and so they need to go into the cookie. I could just use 2 fixed values/struct fields, but was not sure of the original intent there.

File: app/post.go (hunk 9 above)
Comment (levb) [reply app/post.go]: Copied this over from `SendEphemeralPost`. I suspect that there are plenty of places in the go code that assume `post.Props != nil`?

File: model/integration_action.go (hunk 10 above)
Comment (levb) [reply model/integration_action.go]: https://github.com/mattermost/mattermost-server/blob/master/model/post.go#L73 for example. Avoids storing this field in the database?

File: plugin/api.go (hunk 11 above)
Comment (levb) [reply plugin/api.go]: This is a good question. We already had a `CreatePost` and a `SendEphemeralPost`, so I just followed the pattern there. Choosing between "extra/redundant" and "inconsistent" - it seemed that inconsistent would be worse here... And it certainly keeps the code of each API simpler. Yet, I am 0/5 here, can go either way.

File: plugin/api.go (hunk 12 above)
Comment (levb) [reply plugin/api.go]: Let's discuss ^^, resolving this thread.

File: model/integration_action.go (hunk 2 above)
Comment (levb) [reply model/integration_action.go]: Removed TeamId from the cookie (it wasn't being set anyway). If we need it in a specific use-case it'd need to be passed in through SendEphemeralPost, which is difficult without breaking the API.

File: app/integration_action.go (hunk 4 above)
Comment (levb) [reply app/integration_action.go]: Well, ended up not so simple... but seems correct now. Have to make sure Context gets created if nil, don't want to do it if there's nothing to save there, etc. Consolidated the logic for the 2 branches though (cookies and DB)

File: model/config.go (hunk 3 above)
Comment (levb) [reply model/config.go]: @hanzei @crspeller asymmetricSigningKey does not seem to be the right thing to use. 0/5 should stick with a separate setting. Let me know what you think.

Also, @crspeller can you please confirm that the current implementation is correct for the clustered use-case, and that I would just need to add this key to admin panel as Jason suggested?

**Comment by levb** [comment]
A quick check in on the remaining issues:
1. Websocket updates seem to be broken in master webapp, https://mattermost.atlassian.net/browse/MM-14184. It is a blocker for this PR as well.
2. Confirm with @crspeller about the encryption key for cluster, add it to the admin_panel
3. @hanzei's question of should we add these new APIs, or add the "ephemeral" path to `Create/Update/Delete` - thoughts?

Anything else?

**Comment by levb** [comment]
@hanzei,
> When clicking on a message button, nothing happens in the UI
I could not reproduce, except for the missing websocket update. It seems the post rendered correctly upon a refresh. Also, the code checking permissions shouldn't have changed at all with this PR.

**Comment by jasonblais** [comment]
@levb Additional items:

4. We should also add the config setting to the System Console UI (though that's probably for the mattermost-webapp PR) https://github.com/mattermost/mattermost-server/pull/10258#discussion_r255303200

5. Is there QA testing you'd like to have before merging? We can do this either on this PR, or the mattermost-webapp PR.

**Comment by levb** [comment]
@jasonblais TY! #4 and #2 are the same AFAICT; For E2E testing, once we have webapp and redux merged, I can spin up a server on this PR, and install /Giphy there. There needs to be a separate test for PostActionCookies in select inputs, which /Giphy does not use, but short of writing a plugin that uses it I am not sure how to test it E2E.

File: model/config.go (hunk 3 above)
Comment (hanzei) [reply model/config.go]: You're right, `ecdsa.PrivateKey` just exposes a `Sign()` method. We could still do something fancy like generating a ephemeral symmetry key, signing it and sending the key and the signature to the client. We would later check the returned signature and then encrypt with the provided key. 0/5 that this is the right way.

Technically elliptic curve public/private keys are able to encrypt data. `crypto/ecdsa` just doens't support this. But I don't know a library that can do this.

There are two reasons why I'm persistent on this and I think this should not be a system setting:
- It adds jet another system setting
- It exposes something security related to the system admin even tough the admin should not care (and can only screw thing up).
- A sign/encrypt function seams to be useful in general.

File: app/post.go (hunk 9 above)
Comment (hanzei) [reply app/post.go]: If this is copied from somewhere, I'm fine with it. I just wanted to know the reasoning.

File: plugin/api.go (hunk 11 above)
Comment (hanzei) [reply plugin/api.go]: For me this is more of a question about "what the developers wants to do".

When the dev creates a post, he cares **what** kind of post that is. Hence it's fine to have `CreatePost` and `SendEphemeralPost`. But for update some kind of post this is different. The dev already decided what kind of post this is and just wants to change some thing in the post. Why should he care, what kind of post this is?
And for deleting a post, he "just wants to remove that post" and doesn't care about the kind.

Sure, it would be more work for the server to figure this out, but I think the complexity on the server part is worth.

**Comment by hanzei** [comment]
> @hanzei,
> 
> > When clicking on a message button, nothing happens in the UI
> 
> I could not reproduce, except for the missing websocket update. It seems the post rendered correctly upon a refresh. Also, the code checking permissions shouldn't have changed at all with this PR.

Could you please try to reproduce the issue with Matterpoll? You only need to change the command type in https://github.com/matterpoll/matterpoll/blob/master/server/plugin/command.go#L77

You surely didn't change the permission checking directly, but I suspect indirectly:
- With "normal" posts, the server gets the channel member for the post and checks  if the permission is correct. This is done in https://github.com/mattermost/mattermost-server/blob/master/app/authorization.go#L67-L73
- For ephemeral post both https://github.com/mattermost/mattermost-server/blob/master/app/authorization.go#L67 and https://github.com/mattermost/mattermost-server/blob/master/app/authorization.go#L75 fail because the post is not stored in the DB. Hence it return the check it https://github.com/mattermost/mattermost-server/blob/master/app/authorization.go#L82. And this check fails for ephemeral messages.

I don't know, why the code is that way. I only know, that it works this way.

File: plugin/api.go (hunk 11 above)
Comment (levb) [reply plugin/api.go]: For an ephemeral post an Update is effectively a full rewrite - the server has no knowledge of the prior post, except for what's in the cookie (and even that is out of context, without material code changes), so it's completely up to the plugin to re-compose the post. If it weren't for the API compatibility issues, I'd have preferred to add a `sendAsUpdate bool` argument to `SendEphemeralPost`.

I completely see the point for `Delete` - but if we have the other 2 methods separated, then maybe it's worth having one for consistency?

@crspeller @jwilander opinions?

**Comment by levb** [comment]
@hanzei I understand what you meant now with permissions. Will look into it.

File: model/config.go (hunk 3 above)
Comment (levb) [reply model/config.go]: I agree that we should try hard not to expose the setting at all, it is just an internally-used 32 bytes that should be cryptographically random, and shared between all servers in a cluster. I am still unsure that piggybacking on PKI is a good idea here.

File: plugin/api.go (hunk 11 above)
Comment (hanzei) [reply plugin/api.go]: Sure, the plugin/integration must provide the whole post. But why can't it use `UpdatePost()`?

Comment (levb) [reply plugin/api.go]: k, let me make the change and see.

**Comment by crspeller** [comment]
For the issues:
2. Maybe we should just generate a SymmetricEncyrptionKey like we do with AsymmetricSigningKey? Then store it in the DB. Then we don't worry about exposing it to the user or synchronizing over HA.
3. I feel like the semantic differences between posts and ephemeral posts are large enough to justify the separate API. I may have preferred to unify the two concepts, but I think that ship has sailed as we already split it up for creation.

**Comment by hanzei** [comment]
2. :+1: on generating a new key, that can also sign.
3. I think it's worth to look at the differences between posts and ephemeral posts, but this might be out of scope of this PR. What do you think of adding the API's for now, declare them experimental and look into this later?

**Comment by levb** [comment]
Update:

> 1.  Websocket updates seem to be broken in master webapp, https://mattermost.atlassian.net/browse/MM-14184. It is a blocker for this PR as well.

*OPEN* Still affecting this PR and other

> 2. ... encryption key for cluster

*DONE* After discussing with @crspeller re-implemented the key generation similar to AsymmetricSigningKey, saved as a system setting to the database, rather than config. No UI/doc needed. See https://github.com/mattermost/mattermost-server/pull/10258/commits/b100b611d6791d276b93f3a865d295e240b3e997

> You do not have the appropriate permissions

*DONE* Fixed. My bad - was using the sysadmin user when testing. See https://github.com/mattermost/mattermost-server/pull/10258/commits/370cbb9c9aab779cfee24265431a3840c2bcbc24

> 3. @hanzei's question of should we add these new APIs, or add the "ephemeral" path to `Create/Update/Delete` - thoughts?

@hanzei as an extra argument, please look at https://github.com/matterpoll/matterpoll/blob/master/server/plugin/api.go#L118 and compare to https://github.com/levb/mattermost-giphy-plugin/blob/master/server/plugin.go#L194. In particular, with ephemeral messages the caller (plugin) must set the Channel ID, otherwise `UpdatePost` won't know. As @crspeller said, "semantic differences between posts and ephemeral posts are large enough". Stick with separate APIs?

Other than this last issue (3) I think I am done with code changes and tests here.

**Comment by jasonblais** [comment]
Regarding 2: To update the `PostActionCookieSecret`, does the admin execute a SQL query? How often do we think that would be (do we think that would be rare/unlikely)?

**Comment by levb** [comment]
@jasonblais you mean if the secret has been compromised and need to be regenerated? Do you know what do we do for AsymmetricSigningKey? I can do exactly the same, didn't see an API to regenerate it but maybe I missed it.

**Comment by jasonblais** [comment]
I don't know what we do for AsymmetricSigningKey, and maybe there's no need to regenerate - @crspeller or @hanzei are probably more equipped to answer.

File: app/post.go [hunk 13]
```diff
@@ -404,8 +404,40 @@ func (a *App) SendEphemeralPost(userId string, post *model.Post) *model.Post {
...
+	post = model.AddPostActionCookies(post, a.PostActionCookieSecret())
+	message.Add("post", post.ToJson())
+	a.Publish(message)
+
+	return post
+}
+
+func (a *App) DeleteEphemeralPost(userId string, post *model.Post) *model.Post {
+	post.Type = model.POST_EPHEMERAL
+	post.UpdateAt = model.GetMillis()
```
Comment (jespino) [inline app/post.go]: Probably not needed, but being strict, shouldn't we set here the `post.DeleteAt` to `model.GetMillis()`?

File: app/post.go [hunk 14]
```diff
@@ -434,7 +434,7 @@ func (a *App) UpdateEphemeralPost(userId string, post *model.Post) *model.Post {
 
 func (a *App) DeleteEphemeralPost(userId string, post *model.Post) *model.Post {
 	post.Type = model.POST_EPHEMERAL
-	post.UpdateAt = model.GetMillis()
+	post.DeleteAt = model.GetMillis()
```
Comment (jespino) [inline app/post.go]: Being a bit picky, both fields are updated to the same when the post is deleted, so the correct code should be

```suggestion
deleteAt = model.GetMillis()
post.UpdateAt = deleteAt
post.DeleteAt = deleteAt
```

Comment (levb) [reply app/post.go]: It worked without it, so I took it out, but sure, I agree it's more explicit.

**Comment by hanzei** [comment]
@jasonblais `postActionCookieSecret` is not exposed to an admin. The server takes care of generating and storing `postActionCookieSecret`. There should be no reason for neither the server nor the admin to regenerate it.

Does this solve your question?

File: app/integration_action.go [hunk 15]
```diff
@@ -31,90 +31,138 @@ import (
...
-			}
+		for key, value := range retain {
+			response.Update.AddProp(key, value)
+		}
+		for _, key := range remove {
+			delete(response.Update.Props, key)
 		}
-		if _, err := a.UpdatePost(response.Update, false); err != nil {
-			return "", err
+		if _, appErr = a.UpdatePost(response.Update, false); appErr != nil {
```
Comment (hanzei) [inline app/integration_action.go]: When the replied post is a ephemeral post, this call fails. Hence integrations are not able to update there posts via `PostActionIntegrationResponse.Update`.

Is this something we just don't want to support? Or not jet?

**Comment by hanzei** [comment]
Re 2: 

>  In particular, with ephemeral messages the caller (plugin) must set the Channel ID, otherwise `UpdatePost` won't know.

That's a good point. I'm on board with having a separate API for now. But I would like discuss in a separate content the difference between post and ephemeral posts. Especially strange is that both are represented by `model.Post`.

I posted above:
>    2\. What do you think of adding the API's for now, declare them experimental and look into this later?

What do you think about marking them as experimental and looking into them before Plugin stable gets released?

File: app/config.go [hunk 16]
```diff
@@ -106,6 +106,65 @@ func (a *App) RemoveConfigListener(id string) {
...
+			return result.Err
+		}
+
+		if err := json.Unmarshal([]byte(result.Data.(*model.System).Value), &secret); err != nil {
+			return err
+		}
+	}
+
+	a.Srv.postActionCookieSecret = secret.Secret
+	// a.regenerateClientConfig() -- not needed here?
```
Comment (crspeller) [inline app/config.go]: Yes we are presumably not sending this secret to the client. :P

**Comment by hanzei** [comment]
@levb @jasonblais What do you think about creating a HW ticket about adding this to the demo plugin?

**Comment by jasonblais** [comment]
@hanzei Let's go for it!

**Comment by hanzei** [comment]
Could some else please take care of creating the ticket?

**Comment by levb** [comment]
k, doing it now
